File: src/rag/MerchantRAG.ts
/**
 * MerchantRAG - RAG module for Merchant application queries.
 *
 * Retrieves information exclusively from the data/merchant/ directory.
 * Handles: product management, seller onboarding, merchant FAQs, store management.
 */

import * as path from "path";
import { BaseRAGModule, RAGResponse } from "./BaseRAGModule";
import { LocalDocumentStore } from "./LocalDocumentStore";

export class MerchantRAG implements BaseRAGModule {
  readonly moduleName = "merchant_rag";
  private readonly store: LocalDocumentStore;

  constructor(dataRootDir: string) {
    const merchantDataDir = path.join(dataRootDir, "merchant");
    this.store = new LocalDocumentStore(merchantDataDir);
    console.log(`[MerchantRAG] Initialized with data from: ${merchantDataDir}`);
  }

  async retrieve(query: string, topK: number = 3): Promise<RAGResponse> {
    const results = this.store.search(query, topK);

    const context = results.length > 0
      ? results
          .map(
            (r, i) =>
              `[Source: ${r.source} | Relevance: ${r.score}]\n${r.content}`
          )
          .join("\n\n")
      : "No relevant information found in the merchant knowledge base.";

    return {
      context,
      results,
      ragModule: this.moduleName,
    };
  }
}

File: src_python/rag/merchant_rag.py
"""
MerchantRAG - RAG module for Merchant application queries.

Retrieves information exclusively from the data/merchant/ directory.
Handles: product management, seller onboarding, merchant FAQs, store management.
"""

from __future__ import annotations

import os

from .base_rag_module import BaseRAGModule, RAGResponse
from .local_document_store import LocalDocumentStore


class MerchantRAG(BaseRAGModule):
    def __init__(self, data_root_dir: str):
        merchant_data_dir = os.path.join(data_root_dir, "merchant")
        self._store = LocalDocumentStore(merchant_data_dir)
        print(f"[MerchantRAG] Initialized with data from: {merchant_data_dir}")

    @property
    def module_name(self) -> str:
        return "merchant_rag"

    async def retrieve(self, query: str, top_k: int = 3) -> RAGResponse:
        results = self._store.search(query, top_k)

        if results:
            context = "\n\n".join(
                f"[Source: {r.source} | Relevance: {r.score}]\n{r.content}"
                for r in results
            )
        else:
            context = "No relevant information found in the merchant knowledge base."

        return RAGResponse(
            context=context,
            results=results,
            rag_module=self.module_name,
        )
